### PR TITLE
fresh implementation of Names

### PR DESCRIPTION
**Pull request by mkeskells**
inherently threadsafe, and could be shared between Runs - but not in this PR

the main motivation is that in our internal compilation we see that the name table is overloaded and has chains up to 500 long, so looked to see if we could make it faster

This is very much a first cut, but it seems to be faster for the simple JMH test attached
NamesBenchmark - looks up names that exist in the table
NamesBenchmarkInsertNew - inserts names from strings ( that have to calculate their hashcode)

```
OLD
[info] Benchmark                           Mode  Cnt       Score      Error  Units
[info] NamesBenchmark.createBoth           avgt   20   82178.994 ▒ 2159.234  ns/op
[info] NamesBenchmark.createTerm           avgt   20   23055.097 ▒ 1140.830  ns/op
[info] NamesBenchmark.createType           avgt   20   35959.945 ▒ 2379.958  ns/op
[info] NamesBenchmarkInsertNew.createBoth  avgt   20  425193.881 ▒ 2876.201  ns/op
[info] NamesBenchmarkInsertNew.createTerm  avgt   20  286452.250 ▒ 3880.833  ns/op
[info] NamesBenchmarkInsertNew.createType  avgt   20  416672.282 ▒ 3482.710  ns/op

NEW 
[info] Benchmark                           Mode  Cnt       Score      Error  Units
[info] NamesBenchmark.createBoth           avgt   20   25057.350 ▒  136.583  ns/op
[info] NamesBenchmark.createTerm           avgt   20   20373.252 ▒ 1058.120  ns/op
[info] NamesBenchmark.createType           avgt   20   20881.439 ▒  453.519  ns/op
[info] NamesBenchmarkInsertNew.createBoth  avgt   20  139702.194 ▒ 3616.623  ns/op
[info] NamesBenchmarkInsertNew.createTerm  avgt   20  134879.891 ▒ 2139.333  ns/op
[info] NamesBenchmarkInsertNew.createType  avgt   20  137091.462 ▒ 5214.178  ns/op
```

I did see some much faster times when using a class instead of a trait , but could not see a way to make that work without changing a lot

I removed some deprecated APIs and proposed some more that could be inlined in the code IMO, but left these so as to ease review

I have not looked at PDO yet, so there is probably more to gain still

### CODE REVIEW

**Comment by retronym** [comment]
The build is failing with an OOME when using the new compiler to compile the library. That's happening super early in the initalization of `Run`:

```
error: java.lang.OutOfMemoryError: Java heap space
	at scala.reflect.ManifestFactory$IntManifest.newArray(Manifest.scala:221)
	at scala.reflect.ManifestFactory$IntManifest.newArray(Manifest.scala:219)
	at scala.tools.nsc.ast.parser.ScannersCommon.createKeywordArray(Scanners.scala:72)
	at scala.tools.nsc.ast.parser.ScannersCommon.createKeywordArray$(Scanners.scala:68)
	at scala.tools.nsc.ast.parser.SyntaxAnalyzer.createKeywordArray(SyntaxAnalyzer.scala:20)
	at scala.tools.nsc.ast.parser.Scanners.$init$(Scanners.scala:1270)
	at scala.tools.nsc.ast.parser.SyntaxAnalyzer.<init>(SyntaxAnalyzer.scala:20)
	at scala.tools.nsc.Global$$anon$3.<init>(Global.scala:464)
	at scala.tools.nsc.Global.syntaxAnalyzer$lzycompute(Global.scala:462)
	at scala.tools.nsc.Global.syntaxAnalyzer(Global.scala:462)
	at scala.tools.nsc.Global.computeInternalPhases(Global.scala:665)
	at scala.tools.nsc.Global.computePhaseDescriptors(Global.scala:712)
	at scala.tools.nsc.Global.phaseDescriptors$lzycompute(Global.scala:719)
	at scala.tools.nsc.Global.phaseDescriptors(Global.scala:719)
	at scala.tools.nsc.Global$Run.<init>(Global.scala:1226)
	at scala.tools.nsc.Driver.doCompile(Driver.scala:47)
	at scala.tools.nsc.MainClass.doCompile(Main.scala:30)
	at scala.tools.nsc.Driver.process(Driver.scala:68)
	at scala.tools.nsc.Main.process(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sbt.internal.inc.RawCompiler.getReporter$1(RawCompiler.scala:50)
	at sbt.internal.inc.RawCompiler.apply(RawCompiler.scala:71)
	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$6(AnalyzingCompiler.scala:350)
	at sbt.internal.inc.AnalyzingCompiler$$$Lambda$5500/1180376203.apply$mcV$sp(Unknown Source)
	at sbt.internal.inc.AnalyzingCompiler$.handleCompilationError$1(AnalyzingCompiler.scala:327)
	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$4(AnalyzingCompiler.scala:346)
	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$4$adapted(AnalyzingCompiler.scala:341)
	at sbt.internal.inc.AnalyzingCompiler$$$Lambda$5498/156164326.apply(Unknown Source)
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:475)
[error] ## Exception when compiling 557 sources to /home/jenkins/workspace/scala-2.13.x-validate-main@3/build/quick/classes/library
```

File: src/reflect/scala/reflect/internal/Names.scala [hunk 1]
```diff
@@ -14,607 +14,291 @@ package scala
...
-            Console.println("Compared '%s' and '%s', one term, one type.%s".format(this, x,
-              if (panic) "  And they contain the same name string!"
-              else ""
-            ))
-          }
-        case _ =>
+    res
+    //same as cache.computeIfAbsent(value, new NameHolder(_)) but faster
+  }
+  private[this] final val cache = new ConcurrentHashMap[String, NameHolder](100000, 0.75F, 1)
```
Comment (hrhino) [inline src/reflect/scala/reflect/internal/Names.scala]: Doubtfully your memory problem but this seems like a very big table to start out with...

File: src/reflect/scala/reflect/internal/Names.scala [hunk 2]
```diff
@@ -14,607 +14,291 @@ package scala
...
-      val sLen = s.length()
-      if (sLen == 1) return i
-      while (i + sLen <= len) {
-        var j = 1
-        while (s.charAt(j) == chrs(index + i + j)) {
-          j += 1
-          if (j == sLen) return i
-        }
-        i = pos(s.charAt(0), i + 1)
+    abstract class AName extends NameApi with CharSequence {
```
Comment (hrhino) [inline src/reflect/scala/reflect/internal/Names.scala]: I worry slightly about each name string becoming 3 objects now; I'm not sure how common it is to convert between term and type names, now that `newTypeName` isn't implemented in terms of `newTermName`.

**Comment by retronym** [comment]
@mkeskells @hrhino As discussed, here's an implementation of name table sharing across Global instances (an idea borrowed the idea from `dotc`.)

https://github.com/retronym/scala/pull/43

> the main motivation is that in our internal compilation we see that the name table is overloaded and has chains up to 500 long,

@hrhino Do you know what identifiers are leading this problem in your compiles? I expect that #5474 would fix it, although it might also be possible to change the compiler or your plugin phases to avoid unnecessary freshening of some names. 

Also, do you have a sense for why linking the encoded/decoded names is a performance win? Maybe that was just a by-product of the badly distributed hashcodes?

**Comment by diesalbla** [comment]
@mkeskells Are you still working on this Pull Request? The results of those benchmarks looked promising.

**Comment by mkeskells** [comment]
> @mkeskells Are you still working on this Pull Request? The results of those benchmarks looked promising.

I would like to get back to it sometime, but working on 2.12 collection performance ATM, and likely for the next month at least. @diesalbla would you like to work on this together, or take this over?
@retronym has made some changes in the interim to make these methods less hot, so need to check to see the benefit, and the results that I quoted may well not still apply

**Comment by dwijnand** [comment]
> @retronym has made some changes in the interim to make these methods less hot, so need to check to see the benefit, and the results that I quoted may well not still apply

Given that, and the time passed, shall we close this PR for now?  We can always reopen it if activity resumes.

**Pull request closed without merging**